base/threadfuncs.jl

```diff
@@ -0,0 +1,54 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads(),
+                    schedule=:fair)
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` tasks spawned by `Threads.@spawn`. This function
+will wait for all internally spawned tasks to complete before returning.
+
+If `schedule` is `:fair`, `Threads.foreach` will spawn a new task for each work
+item, allowing Julia's scheduler to more freely load-balance work items across
+threads. Technically, in `:fair` mode, `Threads.foreach` may execute up to
+`2 * ntasks` tasks concurrently.
+
+If `schedule` is `:static`, `Threads.foreach` will simply spawn `ntasks` tasks
+such that each task is locked to the thread it was spawned on. This approach has
+lower scheduling overhead and thus may be more suitable for more uniform,
+fine-grained workloads.
+
+!!! compat "Julia 1.6"
+    This function requires Julia 1.6 or later.
+"""
+function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads(),
+                         schedule=:fair)
```

Comment:
If the keyword is not constant propagated `apply` will be type-instable. So this relies on inlining...

Probably better to have schedule be a "trait" so that any instability is pushed to the caller

Reply:
I thought it won't matter match as `Task` needs dynamic dispatch anyway?

Reply:
Yeah, I figured dispatch/spawn overhead would be high enough here that it wouldn't matter, but could definitely be wrong. Easy to add some `FairSchedule`, `StaticSchedule`, etc. traits to `Threads` if it's indeed useful. API-wise, it seems like we should ensure there's parity with the `:static` argument to `@threads`, but could also just use a singleton to make them look the same to users e.g. `const static = StaticSchedule()`. Then make this function signature `Threads.foreach(f, channel::Channel, ::AbstractSchedule)` and call it like `Threads.foreach(f, channel, Threads.static)`...actually I kinda hate that now that I've written it out lol; seems more transparent/readable just to stick with `Threads.foreach(f, channel, Threads.StaticSchedule())`.

Anyway, could add those traits here if useful.

Reply:
I guess for the static schedule, this per-item dispatch overhead could accumulate a bit over the course of the loop. I'll implement these traits, then.

Reply:
Schedule selection is now accomplished via traits instead of `Symbol`s